For the proposed analysis questions we need to populate the below datapoints from the raw data provided. 

| Needed Datapoints |
|---|
| Accident ID |
|Date|
|Time|
|Day/Night|
|State|

The point of this discovery is to determine what can be pulled from the raw dataset (US_Accidents_March23) to produce the datapoints listed in the table above. We will be looking at the format the data is currently in as well as checking for null values in the rows.

In [7]:
# Import needed modules
import pandas as pd

# Load raw accident excel file into variable 'accidents'
accidents = pd.read_csv(r'Data\Raw\US_Accidents_March23.csv')

### Column Name Review
Looking through dataframe info (coding cell below) we can identify numerous column names that are not needed to visualize the proposed data analysis. Columns of interest at first glance to be used: Accident ID, Start_Time, Sunrise_Sunset, and State.


In [ ]:
# Checking column designations
accidents.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7728394 entries, 0 to 7728393
Data columns (total 46 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   ID                     object 
 1   Source                 object 
 2   Severity               int64  
 3   Start_Time             object 
 4   End_Time               object 
 5   Start_Lat              float64
 6   Start_Lng              float64
 7   End_Lat                float64
 8   End_Lng                float64
 9   Distance(mi)           float64
 10  Description            object 
 11  Street                 object 
 12  City                   object 
 13  County                 object 
 14  State                  object 
 15  Zipcode                object 
 16  Country                object 
 17  Timezone               object 
 18  Airport_Code           object 
 19  Weather_Timestamp      object 
 20  Temperature(F)         float64
 21  Wind_Chill(F)          float64
 22  Humidity(%)       

### Reviewing 'ID' Column
The below code cells are looking at the data in the colum 'ID'. This column displays a designation for the row entry.
- There are a total of 7728394 rows in the column
- Data populates as A-'designation'
- Total rows with non-null values is 7728394 so no need to drop null values


In [ ]:
# Checking format of data
accidents['ID'] 

0                A-1
1                A-2
2                A-3
3                A-4
4                A-5
             ...    
7728389    A-7777757
7728390    A-7777758
7728391    A-7777759
7728392    A-7777760
7728393    A-7777761
Name: ID, Length: 7728394, dtype: object

In [ ]:
# Checking for null values
accidents['ID'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: ID
Non-Null Count    Dtype 
--------------    ----- 
7728394 non-null  object
dtypes: object(1)
memory usage: 59.0+ MB


### Reviewing 'Start_Time' Column
The below code cells are looking at the data in the column 'Start_Time'. This column displays the date and start time of the accident. 
- There are a total of 7,728,394 rows in the column
- The date populates in yyyy-mm-dd format and the time is in hh:mm:ss. 
    - Date format used in 0LegalizationDataPull.py is mm/dd/yyyy. We will need to reformat the display of date data in our cleaning script.
    - Seconds not really pertinent for our use, we will drop this in the final formatting.
- We will separate these data points in to individual columns in the final dataframe.
- Total rows with non-null values 7,728,394 so we won't need to drop any rows.


In [ ]:
# Checking format of data
accidents['Start_Time']

0          2016-02-08 05:46:00
1          2016-02-08 06:07:59
2          2016-02-08 06:49:27
3          2016-02-08 07:23:34
4          2016-02-08 07:39:07
                  ...         
7728389    2019-08-23 18:03:25
7728390    2019-08-23 19:11:30
7728391    2019-08-23 19:00:21
7728392    2019-08-23 19:00:21
7728393    2019-08-23 18:52:06
Name: Start_Time, Length: 7728394, dtype: object

In [ ]:
# Checking for null values
accidents['Start_Time'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: Start_Time
Non-Null Count    Dtype 
--------------    ----- 
7728394 non-null  object
dtypes: object(1)
memory usage: 59.0+ MB


### Reviewing the 'Sunrise_Sunset' column
The below code cells are looking at the data in the column 'Sunrise_Sunset'. This column displays whether the accident occurred during the day or night, determined by Sunrise and Sunset time. 
- There are a total of 7,728,394 rows in the column
- The column name doesn't really indicate the stored data, we will rename this column in the final dataframe. The original data used multiple ways to determine day or night (i.e. Nautical, Astronomical, and Civil). We will simply use the data based off of the rise and set of the sun. 
- The date populates a single word of either Day or Night. 
- Total rows with non-null values 7,705,148, so we will have to drop 23,246 rows.

In [ ]:
# Checking format of data
accidents['Sunrise_Sunset']

0          Night
1          Night
2          Night
3          Night
4            Day
           ...  
7728389      Day
7728390      Day
7728391      Day
7728392      Day
7728393      Day
Name: Sunrise_Sunset, Length: 7728394, dtype: object

In [ ]:
# Checking for null values
accidents['Sunrise_Sunset'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: Sunrise_Sunset
Non-Null Count    Dtype 
--------------    ----- 
7705148 non-null  object
dtypes: object(1)
memory usage: 59.0+ MB


### Reviewing the 'State' column
The below code cells are looking at the data in the column 'State'. This column displays the state the accident occurred in. 
- There are a total of 7,728,394 rows in the column
- The date populates an abbreviated entry of the name of the state.
    - The data pulled with 0LegalizationDataPull.py is listed as full state names. We will have to convert those to abbreviations to match this dataset 
- Data was only tracked for the continental US
    - When running the pandas join in 1Cleaning.py we will drop all rows with a designation of HI or AK in the State column from Legalization.csv.
- Total rows with non-null values 7,728,394 so we won't need to drop any rows.

In [ ]:
#Checking the format of the data
accidents['State']

0          OH
1          OH
2          OH
3          OH
4          OH
           ..
7728389    CA
7728390    CA
7728391    CA
7728392    CA
7728393    CA
Name: State, Length: 7728394, dtype: object

In [ ]:
# Checking for the number of each value present in the state column
accidents['State'].value_counts()

State
CA    1741433
FL     880192
TX     582837
SC     382557
NY     347960
NC     338199
VA     303301
PA     296620
MN     192084
OR     179660
AZ     170609
GA     169234
IL     168958
TN     167388
MI     162191
LA     149701
NJ     140719
MD     140417
OH     118115
WA     108221
AL     101044
UT      97079
CO      90885
OK      83647
MO      77323
CT      71005
IN      67224
MA      61996
WI      34688
KY      32254
NE      28870
MT      28496
IA      26307
AR      22780
NV      21665
KS      20992
DC      18630
RI      16971
MS      15181
DE      14097
WV      13793
ID      11376
NM      10325
NH      10213
WY       3757
ND       3487
ME       2698
VT        926
SD        289
Name: count, dtype: int64

In [ ]:
# Checking the count of values in State column
accidents['State'].value_counts().count()

np.int64(49)

In [ ]:
# Checking for null values
accidents['State'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 7728394 entries, 0 to 7728393
Series name: State
Non-Null Count    Dtype 
--------------    ----- 
7728394 non-null  object
dtypes: object(1)
memory usage: 59.0+ MB


### Final Data Discovery Table
Below is the final data discovery table pulled from the discovery process.

The following table lists needed datapoints and potential columns that data can be pulled from.
|Needed Datapoint | Raw column to use | Notes | Raw Datatype |
|---|---|---|---|
| Accident ID | ID | Numeric identifier for accident | str |
|Date| Start_Time | Start_Time has date and time of accident (yyyy-mm-dd hh:mm:ss) | str |
|Time| Start_TIme | Start_Time has date and time of accident | str |
|Day/Night| Sunrise_Sunset | Lists whether it was 'Day' or 'Night' based off of sunrise/sunset | str |
|State| State | This is the name of a state | str |